plugins/lm/lm-nvme.c

```diff
@@ -0,0 +1,988 @@
+		__u8 seqind;
+		__u8 csuuidi;
+		__u8 csvi;
+		__u16 cntlid;
+		__u64 offset;
+		__u32 numd;
+		char  *input;
+		bool  delete;
+	};
+
+	struct config cfg = {
+		.sel = -1,
+		.stype = 0,
+		.seqind = 0,
+		.csuuidi = 0,
+		.csvi = 0,
+		.cntlid = 0,
+		.offset = 0,
+		.numd = 0,
+		.input = NULL,
+		.delete = false
+	};
+
+	OPT_ARGS(opts) = {
+		OPT_BYTE("select",		's', &cfg.sel, select),
+		OPT_SHRT("cntlid",		'c', &cfg.cntlid, cntlid),
+		OPT_BYTE("suspend-type",	't', &cfg.stype, stype),
+		OPT_FLAG("delete",		'd',   &cfg.delete, delete),
+		OPT_BYTE("seq-ind",		'S', &cfg.seqind, seqind),
+		OPT_BYTE("uuid-index",		'U', &cfg.csuuidi, csuuidi),
+		OPT_BYTE("version-index",	'V', &cfg.csvi, csvi),
+		OPT_LONG("offset",		'o', &cfg.offset, offset),
+		OPT_UINT("numd",		'n', &cfg.numd, numd),
+		OPT_FILE("input-file",		'f', &cfg.input, input),
+		OPT_END()
+	};
+
+	err = parse_and_open(&dev, argc, argv, desc, opts);
+	if (err)
+		return err;
+
+	if (cfg.sel == -1) {
+		nvme_show_error("Select field required");
+		return -EINVAL;
+	}
+
+	// Sanity check input parameters
+	if (cfg.sel == MIGRATION_SEND_SUSPEND || cfg.sel == MIGRATION_SEND_RESUME) {
+		if (cfg.csuuidi != 0 || cfg.csvi != 0) {
+			nvme_show_error("Unexpected fields for %s",
+					ARGSTR(lm_migration_send_select_argstr, cfg.sel));
+			return -EINVAL;
+		}
+	} else if (cfg.sel == MIGRATION_SEND_SET_CONTROLLER_STATE) {
+		if (cfg.delete || cfg.stype != 0) {
+			nvme_show_error("Unexpected fields for %s",
+					ARGSTR(lm_migration_send_select_argstr, cfg.sel));
+			return -EINVAL;
+		} else if (!strlen(cfg.input)) {
+			nvme_show_error("Expected file for %s",
+					ARGSTR(lm_migration_send_select_argstr, cfg.sel));
+			return -EINVAL;
+		}
+	}
+
+	if (cfg.input && strlen(cfg.input)) {
+		file = fopen(cfg.input, "r");
+		if (file == NULL) {
+			nvme_show_perror(cfg.input);
+			return -EINVAL;
+		}
+
+		data = nvme_alloc_huge(cfg.numd << 2, &mh);
+		if (!data)
+			return -ENOMEM;
+
+		size_t n_data = fread(data, 1, cfg.numd << 2, file);
+
+		fclose(file);
+
+		if (n_data != (size_t)(cfg.numd << 2)) {
+			nvme_show_error("failed to read controller state data %s", strerror(errno));
+			return -errno;
+		}
+	}
+
+	struct nvme_passthru_cmd cmd = {
+		.opcode = NVME_ADMIN_MIGRATION_SEND,
+		.cdw10 = (cfg.seqind << 16) | cfg.sel,
+		.cdw11 = (cfg.sel == MIGRATION_SEND_SET_CONTROLLER_STATE) ?
+			 (cfg.csuuidi << 24) | (cfg.csvi << 16) | (cfg.cntlid) :
+			 (cfg.delete << 31) | (cfg.stype << 16) | (cfg.cntlid),
+		.cdw12 = (__u32)(cfg.offset),
+		.cdw13 = (__u32)(cfg.offset >> 32),
+		.cdw15 = cfg.numd,
+		.addr =  (unsigned long long)data,
+		.data_len = cfg.numd << 2
+	};
+
+	err = nvme_submit_admin_passthru(dev_fd(dev), &cmd, NULL);
+	if (!err)
+		printf("Migration Send (%s) Successful\n",
+		       ARGSTR(lm_migration_send_select_argstr, cfg.sel));
+	else if (err > 0)
+		nvme_show_status(err);
+	else
+		nvme_show_error("ERROR: nvme_submit_admin_passthru() failed %s", strerror(errno));
```

Comment:
I think this part should be in libnvme like the other helpers.

Reply:
and with this in mind, the same comment is for the other places above. This is in the base spec, so we should also properly support it in the library and not have all the logic in the frontend.